homeassistant/components/group.py

```diff
@@ -179,29 +245,116 @@ def async_setup(hass, config):
     )
 
     @asyncio.coroutine
-    def reload_service_handler(service_call):
+    def reload_service_handler(service):
         """Remove all groups and load new ones from config."""
         conf = yield from component.async_prepare_reload()
         if conf is None:
             return
         yield from _async_process_config(hass, conf, component)
 
+    hass.services.async_register(
+        DOMAIN, SERVICE_RELOAD, reload_service_handler,
+        descriptions[DOMAIN][SERVICE_RELOAD], schema=RELOAD_SERVICE_SCHEMA)
+
+    @asyncio.coroutine
+    def groups_service_handler(service):
+        """Handle dynamic group service functions."""
+        object_id = service.data[ATTR_OBJECT_ID]
+
+        # new group
+        if service.service == SERVICE_SET and object_id not in service_groups:
+            entity_ids = service.data.get(ATTR_ENTITIES) or \
+                service.data.get(ATTR_DELTA) or None
+
+            extra_arg = {attr: service.data[attr] for attr in (
+                ATTR_VISIBLE, ATTR_ICON, ATTR_VIEW, ATTR_CONTROL
+            ) if service.data.get(attr) is not None}
+
+            new_group = yield from Group.async_create_group(
+                hass, service.data.get(ATTR_NAME, object_id),
+                object_id=object_id,
+                entity_ids=entity_ids,
+                user_defined=False,
+                **extra_arg
+            )
+
+            service_groups[object_id] = new_group
+            return
+
+        # update group
+        if service.service == SERVICE_SET:
+            group = service_groups[object_id]
+            need_update = False
+
+            if ATTR_DELTA in service.data:
```

Comment:
When I mentioned delta, I meant what you're doing right now in the code already: only update a property if it is available in the service call.